instance/instance_upgrade.py

```diff
@@ -0,0 +1,159 @@
+        """
+        return OpenEdXInstance.objects.filter(  # Select instances
+            active_appserver___status="running",  # that are running
+            openedx_release__contains=self.INITIAL_RELEASE,  # on {INITIAL_INSTANCE}
+            use_ephemeral_databases=False,  # and use persistent databases.
+        )
+
+    def upgrade_instance(self, instance):
+        """
+        Upgrades single instance. Should not call instance.save() as it is handled in caller.
+        """
+        raise NotImplementedError("Must be overridden in descendant class")
+
+    @classmethod
+    def _set_git_specs(cls, instance):
+        """
+        Updates instance's edx-platform and configuration repositories data (URL and revision)
+        """
+        instance.edx_platform_repository_url = cls.EDX_PLATFORM_SPEC.url
+        instance.edx_platform_commit = cls.EDX_PLATFORM_SPEC.revision
+
+        instance.configuration_source_repo_url = cls.CONFIGURATION_SPEC.url
+        instance.configuration_version = cls.CONFIGURATION_SPEC.revision
+
+    @staticmethod
+    def _replace_extra_config(instance, target_string, new_value):
+        """
+        Replaces part of configuration_extra_settings with other string.
+        """
+        instance.configuration_extra_settings = instance.configuration_extra_settings.replace(target_string, new_value)
+
+    @classmethod
+    def clean_up_after_upgrade(cls, instances):
+        """
+        Should be invoked after all instances are updated
+        """
+        pass
+
+    def upgrade_instances(self):
+        """
+        Main upgrade method:
+        1. Obtains list of instances to upgrade
+        2. Updates instances' fields
+        3. Saves updated instances to DB
+        4. Schedules jobs to spawn new appservers with new instance settings
+        """
+        instances = self.get_instances_to_upgrade()
+
+        for instance in instances:
+            instance.refresh_from_db()
+            logger.info("Upgrading instance %s to %s ...", instance, self.TARGET_RELEASE)
+
+            self.upgrade_instance(instance)
+
+            instance.save()
+            spawn_appserver(instance.ref.pk, mark_active_on_success=True, num_attempts=1)
+
+        # TODO: schedule clean_up_after_upgrade after instances are updated (huey task with conditional?)
+
+
+class DogwoodToEucalyptus1(InstanceUpgrade):
+    """
+    Upgrade from Dogwood to Eucalyptus.1
+    """
+    INITIAL_RELEASE = "dogwood"
+    REVISION = "opencraft-release/eucalyptus.1"
+    TARGET_RELEASE = REVISION
+
+    EDX_PLATFORM_SPEC = GitVersionSpec("https://github.com/open-craft/edx-platform", REVISION)
+    CONFIGURATION_SPEC = GitVersionSpec("https://github.com/open-craft/configuration", REVISION)
+
+    def upgrade_instance(self, instance):
+        """Update instance fields to Eucalyptus.1 values"""
+        self._set_git_specs(instance)
+        instance.configuration_extra_settings += "\nCOMMON_EUCALYPTUS_UPGRADE: true\n"
```

Comment:
@itsjeyd The idea is to add a new line with a setting, so I'm not quite sure how do I use "replace" to add something, and not turn any newline into this new thing. Can you share the algorithm? (you may consider it a programming challenge :))